src/ringmesh/io/geomodel_builder_resqml.cpp

```diff
@@ -0,0 +1,994 @@
+                }
+                std::vector< gmge_id > result;
+                std::set_intersection( horizons_list.begin(),
+                    horizons_list.end(), cur_horizons_list.begin(),
+                    cur_horizons_list.end(), std::back_inserter( result ) );
+                if( result.size() == horizons_list.size() )
+                {
+                    auto id = interp_2_geo_entity_.find( unit.first );
+                    ringmesh_assert( id != interp_2_geo_entity_.end() );
+                    return id->second;
+                }
+            }
+        }
+        return gmge_id();
+    }
+
+    void GeoModelBuilderRESQMLImpl::read_rank_interpreation(
+        StratigraphicColumnRankInterpretation* rank_interp )
+    {
+        showAllMetadata( rank_interp );
+
+        for( auto unit : rank_interp->getStratigraphicUnitInterpretationSet() )
+        {
+            const gmge_id layer = builder_.geology.create_geological_entity(
+                Layer3D::type_name_static() );
+
+            builder_.geology.set_geological_entity_geol_feature(
+                layer, GMGE::GEOL_FEATURE::STRATI_UNIT );
+
+            interp_2_geo_entity_[unit] = layer;
+
+            unit_2_info_[unit].layer_ = layer;
+            if( unit->getInterpretedFeature() != nullptr )
+            {
+                const Layer3D& layer3d =
+                    static_cast< const Layer3D& >( geomodel_.geological_entity(
+                        GeologicalEntityType( "Layer" ), layer.index() ) );
+                builder_.geology.set_geological_entity_name(
+                    layer, unit->getInterpretedFeature()->getTitle() );
+                unit_2_info_[unit].name_ =
+                    unit->getInterpretedFeature()->getTitle();
+            }
+        }
+
+        for( auto contactIndex : range( rank_interp->getContactCount() ) )
+        {
+            unit_2_info_[rank_interp->getSubjectOfContact( contactIndex )]
+                .relation_base_ = get_contact_mode(
+                rank_interp->getSubjectContactModeOfContact( contactIndex ) );
+            unit_2_info_[rank_interp->getSubjectOfContact( contactIndex )]
+                .interface_base_ = interp_2_geo_entity_
+                [(AbstractFeatureInterpretation*) rank_interp
+                        ->getHorizonInterpretationOfContact( contactIndex )];
+
+            unit_2_info_[rank_interp->getDirectObjectOfContact( contactIndex )]
+                .relation_top_ = get_contact_mode(
+                rank_interp->getDirectObjectContactModeOfContact(
+                    contactIndex ) );
+            unit_2_info_[rank_interp->getDirectObjectOfContact( contactIndex )]
+                .interface_top_ = interp_2_geo_entity_
+                [(AbstractFeatureInterpretation*) rank_interp
+                        ->getHorizonInterpretationOfContact( contactIndex )];
+        }
+    }
+
+    void GeoModelBuilderRESQMLImpl::read_strati_column( EpcDocument& pck )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 27 to the 25 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3776)